src/main/java/io/r2dbc/postgresql/codec/AbstractArrayCodec.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2017-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.r2dbc.postgresql.codec;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.r2dbc.postgresql.client.Parameter;
+import io.r2dbc.postgresql.message.Format;
+
+import java.util.Objects;
+import java.util.stream.Stream;
+
+@SuppressWarnings("rawtypes")
+abstract class AbstractArrayCodec<T> extends AbstractCodec<T[]> {
+
+    private final ByteBufAllocator byteBufAllocator;
+
+    AbstractArrayCodec(ByteBufAllocator byteBufAllocator, Class<T[]> type) {
+        super(type);
+        this.byteBufAllocator = Objects.requireNonNull(byteBufAllocator, "byteBufAllocator must not be null");
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    T[] doDecode(ByteBuf byteBuf, Format format, Class<? extends T[]> type) {
+        Objects.requireNonNull(byteBuf, "byteBuf must not be null");
+
+        Stream.Builder<T> builder = Stream.builder();
```

Comment:
Is there a particular reason that you chose the `Stream` API here?  We're generally getting away from it in performance hotspots due to some identified performance issues.

Reply:
Not in particular, I was trying to avoid array copying and I noticed the stream builder uses java.util.stream.SpinedBuffer which intends to do just that. Entirely possible that just using ArrayList would be faster in most cases though.

Reply:
I think I'm going with the devil we know and `ArrayList`.  Let this comment serve as evidence that I'm absolutely not confident in my choice.